File: 2370-longest-ideal-subsequence/2370-longest-ideal-subsequence.java
class Solution {
    Integer dp[][];
    int fun(int ind, char prev, String s, int k)
    {
        if(ind == s.length()) return 0;
        if(dp[ind][prev] != null) return dp[ind][prev];
        int le = 0, nale = 0;
        if(prev == '$' || Math.abs(s.charAt(ind) - prev) <= k)
            le = 1 + fun(ind+1, s.charAt(ind), s, k);
        nale = fun(ind+1, prev, s, k);
        return dp[ind][prev] = Math.max(le, nale);
        
    }
    public int longestIdealString(String s, int k) {
        dp = new Integer[s.length()][130];
        return fun(0, '$', s, k);
    }
}

File: 2370-longest-ideal-subsequence/2370-longest-ideal-subsequence.py
class Solution:
    def __init__(self):
        self.dp = None

    def fun(self, ind, prev, s, k):
        if ind == len(s):
            return 0
        if self.dp[ind][ord(prev)] is not None:
            return self.dp[ind][ord(prev)]
        
        le = 0
        nale = 0
        
        if prev == '$' or abs(ord(s[ind]) - ord(prev)) <= k:
            le = 1 + self.fun(ind + 1, s[ind], s, k)
        nale = self.fun(ind + 1, prev, s, k)
        
        self.dp[ind][ord(prev)] = max(le, nale)
        return self.dp[ind][ord(prev)]

    def longestIdealString(self, s: str, k: int) -> int:
        self.dp = [[None] * 130 for _ in range(len(s))]
        return self.fun(0, '$', s, k)
